src/hyperlight_host/src/hypervisor/mod.rs

```diff
@@ -484,92 +526,303 @@ pub trait InterruptHandle: Debug + Send + Sync {
 
     /// Returns true if the corresponding sandbox has been dropped
     fn dropped(&self) -> bool;
+
+    /// Increment the generation counter for a new guest function call (Linux only).
+    ///
+    /// This must be called exactly once at the start of each guest function call,
+    /// before any VCPU execution begins. The returned generation number will be
+    /// used throughout the entire guest call, even if the VCPU is run multiple
+    /// times (due to host function calls, retries, etc.).
+    ///
+    /// # Returns
+    ///
+    /// The new generation number assigned to this guest function call.
+    ///
+    /// # Note
+    ///
+    /// This is only called on Linux (KVM/MSHV). Windows uses a different interrupt mechanism.
+    #[cfg(any(kvm, mshv))]
+    fn increment_call_generation(&self) -> u64;
+
+    /// Mark that a guest function call is starting.
+    ///
+    /// Increments the generation counter and sets the call_active flag to true,
+    /// indicating that a guest function call is now in progress. This allows
+    /// kill() to stamp cancel_requested with the correct generation.
+    ///
+    /// Must be called at the start of call_guest_function_by_name_no_reset(),
+    /// before any VCPU execution begins.
+    fn set_call_active(&self);
+
+    /// Mark that a guest function call has completed.
+    ///
+    /// Clears the call_active flag, indicating that no guest function call is
+    /// in progress. After this, kill() will have no effect and will return false.
+    ///
+    /// Must be called at the end of call_guest_function_by_name_no_reset(),
+    /// after the guest call has fully completed (whether successfully or with error).
+    fn clear_call_active(&self);
+
+    /// Returns the call_active atomic bool reference for default implementations.
+    ///
+    /// This is used by default trait methods to access the common call_active field.
+    fn get_call_active(&self) -> &std::sync::atomic::AtomicBool;
+
+    /// Returns the dropped atomic bool reference for default implementations.
+    ///
+    /// This is used by default trait methods to access the common dropped field.
+    fn get_dropped(&self) -> &std::sync::atomic::AtomicBool;
+
+    /// Internal method to increment the generation counter.
+    ///
+    /// Returns the new generation value after incrementing.
+    #[cfg(any(kvm, mshv, target_os = "windows"))]
+    fn increment_generation_internal(&self) -> u64;
+}
+
+// Default implementations for common methods
+impl dyn InterruptHandle {
```

Comment:
I believe you can put default method directly in the trait definition, no need to `impl dyn InterruptHandle`